### PR TITLE
feat: add support for Optimistic Concurrency Control

### PR DESCRIPTION
**Pull request by ko3a4ok**
Initially, it was proposed in https://github.com/googleapis/java-spanner/pull/1957

BEGIN_COMMIT_OVERRIDE
chore: add support for Optimistic Concurrency Control
END_COMMIT_OVERRIDE

### CODE REVIEW

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Options.java [hunk 1]
```diff
@@ -299,6 +320,7 @@ void appendToOptions(Options options) {
   private String tag;
   private String etag;
   private Boolean validateOnly;
+  private boolean withOptimisticLock;
```
Comment (rajatbhatta) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/Options.java:323]:
```suggestion
  private Boolean withOptimisticLock;
```

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Options.java [hunk 2]
```diff
@@ -387,6 +409,10 @@ Boolean validateOnly() {
     return validateOnly;
   }
 
+  boolean withOptimisticLock() {
```
Comment (rajatbhatta) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/Options.java:412]:
```suggestion
  Boolean withOptimisticLock() {
```

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Options.java [hunk 3]
```diff
@@ -492,6 +522,9 @@ public int hashCode() {
     if (validateOnly != null) {
       result = 31 * result + validateOnly.hashCode();
     }
+    if (withOptimisticLock) {
+      result = 31 * result + Boolean.hashCode(withOptimisticLock);
```
Comment (rajatbhatta) [inline google-cloud-spanner/src/main/java/com/google/cloud/spanner/Options.java:526]:
```suggestion
      result = 31 * result + withOptimisticLock.hashCode();
```

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/ReadWriteTransactionWithInlineBeginTest.java [hunk 4]
```diff
@@ -389,6 +404,73 @@ public void failedBatchUpdateAndThenUpdate() {
...
+                  assertEquals(rs.getLong(0), 1);
+                }
+              }
+              return null;
+            });
+    Collection<AbstractMessage> requests =
+        Collections2.filter(
+            mockSpanner.getRequests(), msg -> msg.getClass().equals(ExecuteSqlRequest.class));
+    assertEquals(requests.size(), 1);
+    ExecuteSqlRequest request = (ExecuteSqlRequest) Iterables.getOnlyElement(requests);
```
Comment (rajatbhatta) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/ReadWriteTransactionWithInlineBeginTest.java:424]: Can we avoid import an additional library Collections2 here, for filtering? 

We can use [streams and lambdas](https://stackoverflow.com/a/1385698) to achieve the same.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/ReadWriteTransactionWithInlineBeginTest.java [hunk 5]
```diff
@@ -389,6 +404,73 @@ public void failedBatchUpdateAndThenUpdate() {
...
+                  assertEquals(rs.getLong(0), 1);
+                }
+              }
+              return null;
+            });
+    Collection<AbstractMessage> requests =
+        Collections2.filter(
+            mockSpanner.getRequests(), msg -> msg.getClass().equals(ReadRequest.class));
+    assertEquals(requests.size(), 1);
+    ReadRequest request = (ReadRequest) Iterables.getOnlyElement(requests);
```
Comment (rajatbhatta) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/ReadWriteTransactionWithInlineBeginTest.java:446]: Same as above.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/ReadWriteTransactionWithInlineBeginTest.java [hunk 6]
```diff
@@ -389,6 +404,73 @@ public void failedBatchUpdateAndThenUpdate() {
...
+                SpannerException e = assertThrows(SpannerException.class, () -> rs.next());
+                assertEquals(ErrorCode.INVALID_ARGUMENT, e.getErrorCode());
+              }
+              return transaction.executeUpdate(UPDATE_STATEMENT);
+            });
+    Collection<AbstractMessage> requests =
+        Collections2.filter(
+            mockSpanner.getRequests(), msg -> msg.getClass().equals(BeginTransactionRequest.class));
+    assertEquals(requests.size(), 1);
+    BeginTransactionRequest request = (BeginTransactionRequest) Iterables.getOnlyElement(requests);
```
Comment (rajatbhatta) [inline google-cloud-spanner/src/test/java/com/google/cloud/spanner/ReadWriteTransactionWithInlineBeginTest.java:470]: Same as above.

File: google-cloud-spanner/src/main/java/com/google/cloud/spanner/Options.java (hunk 3 above)
Comment (ko3a4ok) [reply google-cloud-spanner/src/main/java/com/google/cloud/spanner/Options.java]: done.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/ReadWriteTransactionWithInlineBeginTest.java (hunk 4 above)
Comment (ko3a4ok) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/ReadWriteTransactionWithInlineBeginTest.java]: sure. done.

File: google-cloud-spanner/src/test/java/com/google/cloud/spanner/ReadWriteTransactionWithInlineBeginTest.java (hunk 6 above)
Comment (ko3a4ok) [reply google-cloud-spanner/src/test/java/com/google/cloud/spanner/ReadWriteTransactionWithInlineBeginTest.java]: done.

**Review by rajatbhatta — APPROVED** [review]
LGTM. Thank you for working on this, @ko3a4ok !

**Pull request merged**